vhdbuilder/packer/install-dependencies.sh

```diff
@@ -395,11 +396,96 @@ for KUBE_PROXY_IMAGE_VERSION in ${KUBE_PROXY_IMAGE_VERSIONS}; do
   CONTAINER_IMAGE="mcr.microsoft.com/oss/kubernetes/kube-proxy:v${KUBE_PROXY_IMAGE_VERSION}"
   pullContainerImage ${cliTool} ${CONTAINER_IMAGE}
   ctr --namespace k8s.io run --rm ${CONTAINER_IMAGE} checkTask /bin/sh -c "iptables --version" | grep -v nf_tables && echo "kube-proxy contains no nf_tables"
-  
+
   # shellcheck disable=SC2181
   echo "  - ${CONTAINER_IMAGE}" >>${VHD_LOGS_FILEPATH}
 done
 
+# download and setup azcopy to use to download private packages with MSI auth
+getAzCopyCurrentPath() {
+  if [[ -f ./azcopy ]]; then
+    echo "./azcopy already exists"
+  else
+    echo "get azcopy at \"${PWD}\"...start"
+    # Download and extract
+    azcopydownloadurl="https://aka.ms/downloadazcopy-v10-linux"
+    if [[ $(isARM64) == 1 ]]; then
+      azcopydownloadurl="https://aka.ms/downloadazcopy-v10-linux-arm64"
+    fi
+    wget "$azcopydownloadurl" -O "downloadazcopy"
+    tar -xvf ./downloadazcopy
+
+    rm -f ./azcopy
+    cp ./azcopy_linux_*/azcopy ./azcopy
+    chmod +x ./azcopy
+
+    rm -f downloadazcopy
+    rm -rf ./azcopy_linux_*/
+    echo "get azcopy...done"
+  fi
+}
+
+# download the given tgz package and extract/import the container image
+importContainerImageTGZ() {
+  CONTAINER_IMAGE_TGZ=$1
+
+  if out=$(ctr --namespace k8s.io image import ${CONTAINER_IMAGE_TGZ}); then # ctr should be there by default now, don't have to handle crictl/docker tools
+    container_image=$(echo $out | cut -d' ' -f 2)
+    echo "imported image $container_image from $CONTAINER_IMAGE_TGZ"
+    retagImageForAllClouds "$CLI_TOOL" "$container_image"
+  else
+    echo "error importing container image from $CONTAINER_IMAGE_TGZ"
+    exit 1
+  fi
+}
+
+# download kubernetes package from the given URL using MSI for auth for azcopy
+# if it is a kube-proxy package, extract image from the downloaded package
+cacheKubePackageFromPrivateUrl() {
+  KUBE_PRIVATE_BINARY_URL=$1
+  IS_KUBE_PROXY_PKG=$2
+
+  echo "process private package url: $KUBE_PRIVATE_BINARY_URL"
+
+  mkdir -p ${K8S_CACHE_DIR} # /opt/kubernetes/downloads/private-packages
+
+  if [[ "$IS_KUBE_PROXY_PKG" == true ]]; then
+    # save kube-proxy pkg with the same/given package name
+    K8S_TGZ_NAME=${KUBE_PRIVATE_BINARY_URL##*/}
+  else
+    # save kube pkg with version number from the url path, this convention is used to find the cached package at run-time
+    K8S_TGZ_NAME=$(echo "$KUBE_PRIVATE_BINARY_URL" | grep -o -P '(?<=\/kubernetes\/).*(?=\/binaries\/)').tar.gz
+  fi
+
+  # use azcopy with MSI instead of curl to download packages
+  getAzCopyCurrentPath
+
+  export AZCOPY_AUTO_LOGIN_TYPE="MSI"
+  export AZCOPY_MSI_RESOURCE_STRING="$LINUX_MSI_RESOURCE_IDS"
+
+  cachedpkg="${K8S_CACHE_DIR}/${K8S_TGZ_NAME}"
+  echo "download private package ${KUBE_PRIVATE_BINARY_URL} and store as ${cachedpkg}"
+  if ! ./azcopy copy "${KUBE_PRIVATE_BINARY_URL}" "${cachedpkg}"; then
```

Comment:
I guess I might be a bit confused: during the VHD build we install the private packages from the private package URL, if supplied, using azcopy. However, at runtime during CSE, if RP provides a private package URL, we download that using curl if the binaries aren't already cached. We couldn't we just avoid azcopy during the build process and use curl like we do at runtime?

Reply:
clarified offline, we should only ever actually be downloading/copying stuff during build-time - CSE side simply checks to see if the VHD _should_ have private packages cached, and if not would fail provisioning

Reply:
*That is how it is implemented, to download packages only at build time. But instead of failing to provision when the packages are not cached, will use the default for that version.